Update README; add clarifying language around DIDs and VCs (#122)
TBD54566975/web5-spec: spec/vc.md

This optionality can be difficult to implement consistently across languages. As a consequence, this specification defines a strict subset of the VC Data Model v1.1 that supports the [plain JSON syntax](https://www.w3.org/TR/vc-data-model/#json), represented by the following table. As a utility JSON Schemas for [Verifiable Credentials](vc-11.json) and [Verifiable Presentations](vp-11.json) are provided to aid in the validation of conformant documents.

## Verifiable Credential Data Model

Following from [this data model](https://www.w3.org/TR/vc-data-model/#basic-concepts).

| Property      | JSON Representation | Required | Notes          |
| ------------- | ------------------- | -------- | -------------- |

> | `@context`    | Array of strings    | Yes      | Contexts defining the meaning of terms within the credential. Must include at least `"https://www.w3.org/2018/credentials/v1"`. |

Reviewer: Well never mind, reading the spec it doesn't have to be the first element

> | `id`          | String              | Yes      | A URI representing a unique identifier for the credential. Recommended to be of form `urn:uuid:3978344f-8596-4c3a-a978-8fcaba3903c5`. |

Reviewer: we will hard not accept this if it is not a URI right?

| `type`        | Array of strings    | Yes      | Type(s) of the credential. Must include `VerifiableCredential`. |

> | `issuer`      | String              | Yes      | A DID representing a unique identifier for the entity that issued the credential. |

Reviewer: will we accept outside vcs that are in this form?

```  
"issuer": {
    "id": "did:example:76e12ec712ebc6f1c221ebfeb1f",
    "name": "Example University"
  },
  ```
  
  and then convert the id internally to just a string?

Author: I need to revise this to be either...I don't think we can get away with supporting just one
the DMV uses an object, but for 90% of our cases we wont

> | `issuanceDate`| String              | Yes      | [XML Datetime](https://www.w3.org/TR/xmlschema11-2/#dateTime) value for when the credential was issued. |

Author: yeah, in practice I've used rfc3339 which is a subset of ISO8601 which is a subset of xml date time

| `expirationDate` | String           | No       | [XML Datetime](https://www.w3.org/TR/xmlschema11-2/#dateTime) value after which the credential is no longer valid. |

> | `credentialSubject` | Object        | Yes      | Data about the subject of the credential. Can be any JSON object. |
> | `credentialSubject.id` | Object     | Yes      | A DID representing a unique identifier for whom the credential's claims are made. |

Reviewer 2: I think you meant...

```suggestion
| `credentialSubject` | Object        | Yes      | Data about the subject of the credential. Can be any JSON object. |
| `credentialSubject.id` | String     | Yes      | A DID representing a unique identifier for whom the credential's claims are made. |
```

Another idea would be to carve out `credentialSubject` into it's own Data Model subheader below, in which case the table would just be of length 1 with `id` but the consistency may be easier to comprehend at first glance (I had to double take what was going on here with `credentialSubject` being the only row in this table with a `.*` property). Either or is fine 👍

Author: yes I'll do that

| `credentialStatus` | Object defined by [Credential Status](#credential-status) | No | Only to be used with [Status List 2021](https://www.w3.org/community/reports/credentials/CG-FINAL-vc-status-list-2021-20230102/). |
| `credentialSchema` | Object defined by [Credential Schema](#credential-schema) | No | Recommended. Only to be used with the type [`JsonSchema`](https://w3c.github.io/vc-json-schema/#jsonschema). |

> | `evidence`    | Object              | No       | Any JSON object as per [Evidence](https://www.w3.org/TR/vc-data-model/#evidence). |

Author: this is the one I was least sure of I know we need evidence, but I'm not sure where.

yes each type field is required

I am open to making this an array of objects for future proofing or alternatively dropping it all together until we have more info wdyt @mistermoe

> | `evidence`    | Object              | No       | Any JSON object as per [Evidence](https://www.w3.org/TR/vc-data-model/#evidence). |

Author: updating this to array and adding a note on not implementing evidence until we have a type we can rely upon

**Additional Notes:**
- The `credentialSubject` property can be any JSON object. It is recommended that this object is defined by an associated `credentialSchema`.
- No [JSON-LD processing](https://www.w3.org/TR/vc-data-model/#json-ld) is performed.

> - Embedded proofs, using the `proof` property must not be present.

Reviewer: if we accept an outside vc with proofs we can still accept it and just drop it?
Or there should be no such thing as a vcjwt with proofs too ?

Author: should be no such thing

> - The `type` property may only contain `VerifiableCredential` or the URI of a JSON Schema, if one is used for the credential.

Reviewer 2: Slight confusion here, in the table above we have `Type(s) of the credential. Must include VerifiableCredential.` but then here we use the word `or` 

I think what you may be describing here is...

```suggestion
- The `type` property must always contain `VerifiableCredential` but may also contain the URI(s) of a JSON Schema, if one is used for the credential.
```

Or do I have that wrong?

Reviewer 2: On a similar note, in two the tables in this doc in one row you have `Must include at least` but then in another row you have `Must include` 

In this table that is: 
- "Contexts defining the meaning of terms within the credential. Must include at least `"https://www.w3.org/2018/credentials/v1"`."
- "Type(s) of the credential. Must include `VerifiableCredential`."

But also in the VP DM below

Maybe make consistent one way or the other

Author: will adjust good spot

> - The `type` property may only contain `VerifiableCredential` or the URI of a JSON Schema, if one is used for the credential.

Reviewer: This is not how we do it currently,

```
      const vc = await VerifiableCredential.create({
        type    : 'TBDeveloperCredential',
        subject : did.uri,
        issuer  : did.uri,
        data    : {
          legit: true
        }
      });
```

So from what you are saying our type "TBDeveloperCredential" cannot be added unless we also include the URI of the json schema is that correct?

If that is a hard rule then we will need to refactor our codebase

Author: yes, this isn't useful IMO and breaks the spec
if we have a need for a concrete type to switch on we should be using a JSON Schema

> - We do not support multiple credential subjects.

Reviewer: huge 👏

- Verifiable Credentials must be secured as JWTs according to the [rules laid out in the specification](https://www.w3.org/TR/vc-data-model/#json-web-token).

### Credential Status Data Model

Following from [this data model](https://www.w3.org/community/reports/credentials/CG-FINAL-vc-status-list-2021-20230102/).

> **StatusList2021Entry**

Reviewer: Are you still going to add BitstringStatusListEntry section?

Author: when we support v2 of the data model I will add an entire section on that and bitstring status list

| `statusListIndex` | String            | Yes      | An integer >= 0 expressed as a string that identifies the bit position of the status of the associated verifiable credential. |
| `statusListCredential` | String       | Yes      | A URL which uniquely identifies a verifiable credential whose type is `StatusList2021Credential`. |

**Additional Notes:**
- When representing Status List Credentials, as opposed to including a status in another VC, the Status List Credential must [follow the guidance here](https://www.w3.org/community/reports/credentials/CG-FINAL-vc-status-list-2021-20230102/#statuslist2021credential).

### Credential Schema Data Model

Following from [this data model](https://w3c.github.io/vc-json-schema/#jsonschema).

| Property        | JSON Representation | Required | Notes          |
| --------------- | ------------------- | -------- | -------------- |
| `id`            | String              | Yes      | A URL which uniquely identifies the JSON Schema for the associated Verifiable Credential. |
| `type`          | String              | Yes      | Must be set to `JsonSchema`. |

**Additional Notes:**

> - Though [the referenced spec](https://w3c.github.io/vc-json-schema/) is designed for v2 of the VC Data Model, we are using it anyway since there is no standard way to use the `credentialSchema` property in v1.1 of the data model.

Reviewer 2: ```suggestion
- Although [the referenced spec](https://w3c.github.io/vc-json-schema/) is designed for v2 of the VC Data Model, we apply it to v1.1 as a standard means to implement the `credentialSchema`.
```

| Property      | JSON Representation | Required | Notes          |
| ------------- | ------------------- | -------- | -------------- |
| `@context`    | Array of strings    | Yes      | Contexts defining the meaning of terms within the presentation. Must include at least `"https://www.w3.org/2018/credentials/v1"`. |
| `id`          | String              | Yes      | A URI representing a unique identifier for the presentation. Recommended to be of form `urn:uuid:3978344f-8596-4c3a-a978-8fcaba3903c5`. |
| `type`        | Array of strings    | Yes      | Type(s) of the presentation. Must include `VerifiablePresentation`. |
| `holder`      | String              | Yes      | A DID representing a unique identifier for the entity that created the presentation. |
| `issuanceDate`| String              | Yes      | [XML Datetime](https://www.w3.org/TR/xmlschema11-2/#dateTime) value for when the presentation was created. |
| `expirationDate` | String           | No       | [XML Datetime](https://www.w3.org/TR/xmlschema11-2/#dateTime) value after which the presentation is no longer valid. |

> | `verifiableCredentials` | Array of strings | Yes    | An array with at least one value, containing the JWT representation of [Verifiable Credential](#verifiable-credential-data-model) objects. |

Reviewer: ```suggestion
| `verifiableCredential` | Array of strings | Yes    | An array with at least one value, containing the JWT representation of [Verifiable Credential](#verifiable-credential-data-model) objects. |
```

**Additional Notes:**
- No [JSON-LD processing](https://www.w3.org/TR/vc-data-model/#json-ld) is performed.
- Embedded proofs, using the `proof` property must not be present.

> - The `type` property may only contain `VerifiablePresentation` or the URI of a JSON Schema, if one is used for the presentation.

Reviewer: ya a bit of confusion:

```
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
    "https://www.w3.org/2018/credentials/examples/v1"
  ],
  "id": "urn:uuid:3978344f-8596-4c3a-a978-8fcaba3903c5",
  "type": ["VerifiablePresentation", "CredentialManagerPresentation"],
  "verifiableCredential": [{  }],
  "proof": [{  }]
}

```

in the spec they have an extra type just added in.

Are you saying we are drawing a hard line and only adding a type if it is a URI of a JSON Schema?

Will we accept creds that just have a random type like the example?

> - The `type` property may only contain `VerifiablePresentation` or the URI of a JSON Schema, if one is used for the presentation.

Author: > Are you saying we are drawing a hard line and only adding a type if it is a URI of a JSON Schema?
this

> Will we accept creds that just have a random type like the example?

it's not random in this example because the term is defined in the provided example context
because we're not doing json ld processing we should not have a type like this

if we use a json schema we can include it here, .... or just include it in the `credentialSchema` property

> - The `type` property may only contain `VerifiablePresentation` or the URI of a JSON Schema, if one is used for the presentation.

Author: updated the language in both places
